worker/rhpv3.go

```diff
@@ -498,111 +442,131 @@ func readSectorCost(pt *rhpv3.HostPriceTable) (types.Currency, error) {
+	if ongoing && time.Now().Before(hpt.Expiry.Add(priceTableValidityLeeway)) {
+		return
+	}
+
+	// price table is being updated, wait for the update
+	if ongoing {
 		select {
 		case <-ctx.Done():
-			return rhpv3.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
-		case <-ongoing.done:
-		}
-		if ongoing.err != nil {
-			return rhpv3.HostPriceTable{}, ongoing.err
-		} else {
-			return *ongoing.pt, nil
+			return hostdb.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
+		case <-update.done:
 		}
+		return update.hpt, update.err
 	}
 
-	// Update price table.
-	var hpt rhpv3.HostPriceTable
-	err := withTransportV3(ctx, siamuxAddr, hk, func(t *rhpv3.Transport) (err error) {
-		hpt, err = RPCPriceTable(t, payFn)
-		return err
-	})
+	// this thread is updating the price table
+	defer func() {
+		update.hpt = hpt
+		update.err = err
+		close(update.done)
 
-	pt.mu.Lock()
-	defer pt.mu.Unlock()
+		p.mu.Lock()
+		if err == nil {
+			p.hpt = hpt
+		}
+		p.update = nil
+		p.mu.Unlock()
+	}()
 
-	// On success we update the pt.
-	if err == nil {
-		ongoing.pt = &hpt
-		pt.pt = &hpt
-		pt.expiry = time.Now().Add(hpt.Validity)
+	// fetch the host info from the bus, it might have a new price table already
+	// in which case we can return without having to update it
+	host, err := b.Host(ctx, hk)
+	if err == nil && host.PriceTable != nil && time.Now().Before(host.PriceTable.Expiry.Add(priceTableValidityLeeway)) {
+		hpt = *host.PriceTable
+		return
+	}
+
+	// try to pay by EA if possible
+	if revision == nil {
+		if cs, err1 := b.ConsensusState(ctx); err1 == nil && cs.Synced {
+			if pt, err2 := w.fetchPriceTable(ctx, hk, host.Settings.SiamuxAddr(), w.preparePriceTableAccountPayment(hk, cs.BlockHeight)); err2 == nil {
+				hpt = pt
+				return
+			}
+		}
 	}
-	// Signal that the update is over.
-	ongoing.err = err
-	close(ongoing.done)
-	pt.ongoingUpdate = nil
-	return hpt, err
-}
 
-// priceTable returns a priceTable from priceTables for the given host or
-// creates a new one.
-func (pts *priceTables) priceTable(hk types.PublicKey) *priceTable {
-	pts.mu.Lock()
-	defer pts.mu.Unlock()
+	// pay by FC if EA payment failed or if a revision was given
+	if revision != nil {
+		return w.fetchPriceTable(ctx, hk, host.Settings.SiamuxAddr(), w.preparePriceTableContractPayment(hk, revision))
+	}
 
-	if _, exists := pts.priceTables[hk]; !exists {
-		pts.priceTables[hk] = &priceTable{hk: hk}
+	// fetch the active contract
+	contract, err := b.ActiveContract(ctx, hk)
+	if err != nil {
+		return hostdb.HostPriceTable{}, err
 	}
-	return pts.priceTables[hk]
+
+	// fetch the price table
+	_ = w.withRevisionV3(ctx, contract.ID, hk, host.Settings.SiamuxAddr(), lockingPriorityPriceTable, lockingDurationPriceTable, func(revision types.FileContractRevision) error {
+		hpt, err = w.fetchPriceTable(ctx, hk, host.Settings.SiamuxAddr(), w.preparePriceTableContractPayment(hk, &revision))
+		return err
+	})
```

Comment:
`withRevisionV3` calls `FetchRevisionWithContract` which calls `priceTable.fetch()` which calls `withRevisionV3` again.

I think we need to detangle this a bit to avoid running into issues in the future. Maybe `withRevisionV3` can pass the pricetable in addition to the revision since it needs to fetch one anyway. Or something similar. Then we don't need to call `fetchPriceTable` again here and can isolate calling it in a single spot.

Reply:
If we don't fall back to contract payments, we don't need this any longer. I considered making this change anyway but I haven't done it yet. I feel passing the price table is a little bit odd since we don't do that for other withX functions neither.